### PR TITLE
Add some guidance on how to model deprecations

### PR DESCRIPTION
**Pull request by ldaley**

### CODE REVIEW

File: subprojects/logging/src/main/java/org/gradle/internal/deprecation/DeprecationLogger.java [hunk 1]
```diff
@@ -31,17 +31,28 @@
  * The basic deprecation message structure is "Summary. DeprecationTimeline. Context. Advice. Documentation."
  *
  * The deprecateX methods in this class return a builder that guides creation of the deprecation message.
+ *
  * Summary is populated by the deprecateX methods in this class.
+ *
  * Context can be added in free text using {@link DeprecationMessageBuilder#withContext(String)}.
+ * It should provide any _instance specific_ details about the deprecated usage that would help a user identify
```
Comment (lptr) [inline subprojects/logging/src/main/java/org/gradle/internal/deprecation/DeprecationLogger.java:38]: Use `<em>` here for emphasis maybe?

File: subprojects/logging/src/main/java/org/gradle/internal/deprecation/DeprecationLogger.java [hunk 2]
```diff
@@ -31,17 +31,28 @@
...
  * - ${@link DeprecationMessageBuilder#willBecomeAnErrorInGradle7()}
+ *
  * After DeprecationTimeline is set, Documentation reference must be added using one of:
  * - {@link DeprecationMessageBuilder.WithDeprecationTimeline#withUpgradeGuideSection(int, String)}
  * - {@link DeprecationMessageBuilder.WithDeprecationTimeline#withDslReference(Class, String)}
  * - {@link DeprecationMessageBuilder.WithDeprecationTimeline#withUserManual(String, String)}
  *
+ * Generally, only one deprecation of a type (i.e. summary/advice) should be emitted during a logical unit of work.
+ * If there are multiple instances of the same logical deprecation during a logical unit of work,
+ * the individual instance specifics should be included as part of the context.
```
Comment (lptr) [inline subprojects/logging/src/main/java/org/gradle/internal/deprecation/DeprecationLogger.java:54]: What is a logical unit of work here? Even if we can't provide a precise definition, an example would be useful, like how does this work with a configuration- or a task-related deprecation.

Note that: we do de-duplication based on the message text, and in the past it was generally accepted that you can produce the same text as many times as needed.

File: subprojects/logging/src/main/java/org/gradle/internal/deprecation/DeprecationLogger.java [hunk 3]
```diff
@@ -31,17 +31,28 @@
  * The basic deprecation message structure is "Summary. DeprecationTimeline. Context. Advice. Documentation."
  *
  * The deprecateX methods in this class return a builder that guides creation of the deprecation message.
+ *
  * Summary is populated by the deprecateX methods in this class.
+ *
  * Context can be added in free text using {@link DeprecationMessageBuilder#withContext(String)}.
+ * It should provide any _instance specific_ details about the deprecated usage that would help a user identify
+ * the full nature of the use site. This is particularly important when the use site is not directly within user code.
```
Comment (lptr) [inline subprojects/logging/src/main/java/org/gradle/internal/deprecation/DeprecationLogger.java:39]: The term "use site" is unclear to me. Is this a well-known term? Do you mean "call site"?

File: subprojects/logging/src/main/java/org/gradle/internal/deprecation/DeprecationLogger.java (hunk 1 above)
Comment (lptr) [reply subprojects/logging/src/main/java/org/gradle/internal/deprecation/DeprecationLogger.java:38]: I think "instance" is not well-defined here. Do you mean the instance the deprecation is related to, or the instance of deprecation?

I think an example with tasks or a configuration could help clarify what you mean here.

File: subprojects/logging/src/main/java/org/gradle/internal/deprecation/DeprecationLogger.java (hunk 3 above)
Comment (ldaley) [reply subprojects/logging/src/main/java/org/gradle/internal/deprecation/DeprecationLogger.java]: A call site is a type of use site. 

Some deprecated features are not “called”. For example, invalid annotation combinations. They are used, and used at a site.

**Comment by ldaley** [comment]
@lptr I did another pass. PTAL.

**Comment by ldaley** [comment]
As I tried to write this, and went looking at examples of deprecation logging it became hard to define rules that match existing patterns. I think there's a good chance that with some more consideration and modelling, we could do better here and provide better service to users. We have anecdotal evidence from Gradle users, including ourselves, that comprehending and processing deprecations can be difficult.

I'm reluctant to spend much more time trying to document how deprecations should be used, when there doesn't actually seem to be a clear model. Or at least, no real motivation or reason to have a clear model at this time.

Build scans use a model of grouping notices by `summary`/`advice`, and allowing use site specific to be conveyed in the `context`. This was done in order to support usage patterns at the time. For example, [here](https://github.com/gradle/gradle/blob/5cd77b034c98c47d6bfa7e532462373902d74373/subprojects/core/src/main/java/org/gradle/api/internal/resources/DefaultTextResourceFactory.java#L94) we didn't want to deprecate specific HTTP URLs, but the use of HTTP in general. Therefore, the `context` carried the URL specific information to allow users to track the use site down.

If I look around the codebase, I can see that this pattern has not been universally adopted. This isn't surprising, as it was never really prescribed or enforced.

**Pull request closed without merging**